File: src/infrastructure/analyzers/micro_tremor_analyzer.py
"""Micro-tremor analyzer — 8-12Hz involuntary facial oscillation detection.

All living humans exhibit involuntary micro-tremor at 8-12Hz caused by
continuous muscle activation (physiological tremor). This signal is:
- Always present in live faces
- Impossible to reproduce on a screen (pixel grid filters it out)
- Too subtle to fake with a held-up photo (hand tremor is 3-5Hz, not 8-12Hz)

This is the strongest single signal for video replay detection because
screen playback physically cannot transmit sub-pixel 8-12Hz oscillation.

Algorithm:
1. Track face centroid from landmarks per frame
2. Accumulate in temporal buffer (90+ frames for 8Hz resolution)
3. Remove low-frequency drift (detrend)
4. FFT on position time-series
5. Measure power in 8-12Hz band relative to total power
6. High relative power = live person
"""

import time
import logging
from collections import deque

import numpy as np

from src.domain.models import FaceROI, AnalyzerResult

logger = logging.getLogger(__name__)

TREMOR_LOW_HZ = 7.0
TREMOR_HIGH_HZ = 13.0
MIN_FRAMES = 45   # ~1.5s at 30fps, ~3s at 15fps
GOOD_FRAMES = 90  # ~3s at 30fps


class MicroTremorAnalyzer:
    """Detects 8-12Hz involuntary micro-tremor from face landmarks.

    Requires landmark data from BlinkAnalyzer via set_landmarks().

    Score:
    - 0-30: No tremor detected (SPOOF — screen/photo/mask)
    - 30-60: Weak tremor (ambiguous)
    - 60-100: Tremor detected (LIVE — real person)
    """

    def __init__(self, fps: float = 30.0):
        self._fps = fps
        self._states: dict[int, deque] = {}
        self._current_landmarks: np.ndarray | None = None
        self._frame_times: deque = deque(maxlen=60)

    @property
    def name(self) -> str:
        return "micro_tremor"

    def set_landmarks(self, landmarks: np.ndarray | None):
        self._current_landmarks = landmarks

    def analyze(self, face_crop: np.ndarray, face_roi: FaceROI) -> AnalyzerResult:
        start = time.perf_counter()

        # Measure actual FPS
        self._frame_times.append(start)
        if len(self._frame_times) > 10:
            dt = self._frame_times[-1] - self._frame_times[0]
            if dt > 0:
                self._fps = (len(self._frame_times) - 1) / dt

        fid = face_roi.face_id
        if fid not in self._states:
            self._states[fid] = deque(maxlen=180)  # 6s buffer

        # Use landmark centroid if available, otherwise bbox center
        if self._current_landmarks is not None and len(self._current_landmarks) > 10:
            cx = float(np.mean(self._current_landmarks[:, 0]))
            cy = float(np.mean(self._current_landmarks[:, 1]))
        else:
            cx, cy = face_roi.bbox.center

        self._states[fid].append((cx, cy))

        elapsed_ms = (time.perf_counter() - start) * 1000

        if len(self._states[fid]) < MIN_FRAMES:
            return AnalyzerResult(
                name=self.name, score=50.0,
                details={"warmup": True, "frames": len(self._states[fid])},
                elapsed_ms=elapsed_ms,
            )

        positions = np.array(list(self._states[fid]))  # (N, 2)
        n = len(positions)

        # Analyze X and Y independently
        results = []
        for axis in range(2):
            signal = positions[:, axis].astype(np.float64)

            # Detrend: remove slow drift (head movement)
            kernel_size = min(15, n // 3)
            if kernel_size > 1:
                smooth = np.convolve(signal, np.ones(kernel_size) / kernel_size, mode="same")
                signal = signal - smooth

            # Hanning window
            window = np.hanning(n)
            signal = signal * window

            # FFT
            fft = np.fft.rfft(signal)
            magnitude = np.abs(fft)
            freqs = np.fft.rfftfreq(n, d=1.0 / self._fps)

            # Tremor band power
            tremor_mask = (freqs >= TREMOR_LOW_HZ) & (freqs <= TREMOR_HIGH_HZ)
            if not np.any(tremor_mask):
                results.append(0.0)
                continue

            tremor_power = float(np.mean(magnitude[tremor_mask]))

            # Total power (excluding DC)
            total_mask = freqs > 0.5
            total_power = float(np.mean(magnitude[total_mask])) if np.any(total_mask) else 1e-6

            # Relative tremor power
            tremor_ratio = tremor_power / max(total_power, 1e-6)
            results.append(tremor_ratio)

        avg_tremor_ratio = sum(results) / max(len(results), 1)
        data_quality = min(1.0, n / GOOD_FRAMES)

        # Score: high tremor ratio = live
        if avg_tremor_ratio > 1.5:
            score = 70.0 + data_quality * 30.0  # Strong tremor
        elif avg_tremor_ratio > 1.0:
            score = 50.0 + avg_tremor_ratio * 15.0
        elif avg_tremor_ratio > 0.5:
            score = 30.0 + avg_tremor_ratio * 20.0
        else:
            # No tremor
            if n > GOOD_FRAMES:
                score = 10.0  # Enough data, no tremor = spoof
            else:
                score = 30.0  # Not enough data yet

        elapsed_ms = (time.perf_counter() - start) * 1000

        return AnalyzerResult(
            name=self.name,
            score=max(0.0, min(100.0, score)),
            details={
                "tremor_ratio": round(avg_tremor_ratio, 4),
                "tremor_x": round(results[0], 4) if len(results) > 0 else 0,
                "tremor_y": round(results[1], 4) if len(results) > 1 else 0,
                "measured_fps": round(self._fps, 1),
                "frames": n,
                "data_quality": round(data_quality, 2),
            },
            elapsed_ms=elapsed_ms,
        )

File: web/src/infrastructure/analyzers/MicroTremorAnalyzer.ts
// Port of src/infrastructure/analyzers/micro_tremor_analyzer.py:1-163
//
// Micro-Tremor — fusion weight 2.5.
//
// Detects 8-12 Hz involuntary physiological tremor that all live humans
// exhibit. Screens cannot transmit this band (pixel grid filters it),
// hand tremor lives at 3-5 Hz so a held photo can't fake it either.
//
// Approximation:
//   * The Python source uses `np.fft.rfft` over ~90-frame windows. We
//     replace it with a hand-rolled DFT — at N≈90 the cost is well
//     under 1 ms (90 × 45 mults). Documented inline.
//   * `np.convolve(..., mode='same')` is replaced with a centered moving
//     average (handle endpoints by clamping). Same shape, same purpose.
//   * `np.hanning` is also hand-rolled.
//
// History buffer: 30 frames per face track (matches WARMUP_FRAMES). The
// Python original allocates 180 (= 6s @ 30fps); the task spec asks for
// 30 — at 30 fps that gives a 1 Hz frequency resolution which is just
// barely enough to resolve the 8-12 Hz band. Documented as a known
// trade-off — callers who need higher precision should pass
// `historyLen: 90` for parity with the Python.

import {
  AnalyzerResult,
  FaceROI,
  IFaceAnalyzer,
  makeAnalyzerResult,
} from "../../domain/models";

const TREMOR_LOW_HZ = 7.0;
const TREMOR_HIGH_HZ = 13.0;
const MIN_FRAMES = 30; // ~1s @ 30fps. Python uses 45.
const GOOD_FRAMES = 90; // ~3s @ 30fps.

export interface MicroTremorOptions {
  /** Fallback fps if not yet measured. Default 30. */
  fps?: number;
  /** Per-track ring buffer length. Default 30 (matches WARMUP_FRAMES). */
  historyLen?: number;
}

export class MicroTremorAnalyzer implements IFaceAnalyzer {
  readonly name = "micro_tremor";

  private fps: number;
  private readonly historyLen: number;

  // Per-face: array of (cx, cy) positions, last `historyLen` entries.
  private states: Map<number, Array<[number, number]>> = new Map();
  // Frame-time ring (last 60). Used to estimate effective fps.
  private frameTimes: number[] = [];
  private currentLandmarks: Float32Array | null = null;

  constructor(options: MicroTremorOptions = {}) {
    this.fps = options.fps ?? 30.0;
    this.historyLen = options.historyLen ?? 30;
  }

  /** Optional API-parity hook: provide the latest landmarks set so we can
   *  use the centroid instead of the bbox center. */
  setLandmarks(landmarks: Float32Array | null): void {
    this.currentLandmarks = landmarks;
  }

  analyze(_faceCrop: ImageData | null, face: FaceROI): AnalyzerResult {
    const start = performance.now();

    // === Measure fps. ===
    this.frameTimes.push(start);
    if (this.frameTimes.length > 60) this.frameTimes.shift();
    if (this.frameTimes.length > 10) {
      const dt =
        (this.frameTimes[this.frameTimes.length - 1] - this.frameTimes[0]) /
        1000.0;
      if (dt > 0) {
        // Real browser frame rates land in 1-120 fps; anything outside
        // means synthetic test loop (too fast) or background-tab throttle
        // (too slow) — both garbage signals for our FFT band math.
        const measured = (this.frameTimes.length - 1) / dt;
        if (measured >= 1 && measured <= 120) this.fps = measured;
      }
    }

    const fid = face.face_id;
    let buf = this.states.get(fid);
    if (!buf) {
      buf = [];
      this.states.set(fid, buf);
    }

    // Centroid: prefer landmarks (pulled off FaceROI or via setLandmarks).
    const lm = face.landmarks ?? this.currentLandmarks;
    let cx: number;
    let cy: number;
    if (lm && lm.length >= 20) {
      let sx = 0;
      let sy = 0;
      const n = Math.floor(lm.length / 2);
      for (let i = 0; i < n; i++) {
        sx += lm[2 * i];
        sy += lm[2 * i + 1];
      }
      cx = sx / n;
      cy = sy / n;
    } else {
      cx = face.bbox.center[0];
      cy = face.bbox.center[1];
    }

    buf.push([cx, cy]);
    if (buf.length > this.historyLen) buf.shift();

    if (buf.length < Math.min(MIN_FRAMES, this.historyLen)) {
      return makeAnalyzerResult(
        this.name,
        50.0,
        { warmup: true, frames: buf.length },
        performance.now() - start,
      );
    }

    const n = buf.length;

    // Analyze X and Y axes independently.
    const ratios: number[] = [];
    for (let axis = 0; axis < 2; axis++) {
      const signal = new Float64Array(n);
      for (let i = 0; i < n; i++) signal[i] = buf[i][axis];
      ratios.push(this.bandPowerRatio(signal));
    }

    const avgRatio = (ratios[0] + ratios[1]) / 2.0;
    const dataQuality = Math.min(1.0, n / GOOD_FRAMES);

    let score: number;
    if (avgRatio > 1.5) {
      score = 70.0 + dataQuality * 30.0;
    } else if (avgRatio > 1.0) {
      score = 50.0 + avgRatio * 15.0;
    } else if (avgRatio > 0.5) {
      score = 30.0 + avgRatio * 20.0;
    } else {
      score = n > GOOD_FRAMES ? 10.0 : 30.0;
    }
    score = Math.max(0.0, Math.min(100.0, score));

    return makeAnalyzerResult(
      this.name,
      score,
      {
        tremor_ratio: round(avgRatio, 4),
        tremor_x: round(ratios[0], 4),
        tremor_y: round(ratios[1], 4),
        measured_fps: round(this.fps, 1),
        frames: n,
        data_quality: round(dataQuality, 2),
      },
      performance.now() - start,
    );
  }

  /**
   * Detrend → window → DFT → tremor-band power / total-power.
   * Returns a unitless ratio; >1 means tremor band has above-average power.
   */
  private bandPowerRatio(signal: Float64Array): number {
    const n = signal.length;
    // === Detrend (Python: signal - moving-average kernel, mode='same'). ===
    const k = Math.min(15, Math.max(1, Math.floor(n / 3)));
    if (k > 1) {
      const ma = movingAverageSame(signal, k);
      for (let i = 0; i < n; i++) signal[i] -= ma[i];
    }

    // === Hanning window (in-place). ===
    for (let i = 0; i < n; i++) {
      // Python np.hanning: 0.5 * (1 - cos(2π i / (N-1))). N==1 → 1.0.
      const w =
        n > 1 ? 0.5 * (1 - Math.cos((2 * Math.PI * i) / (n - 1))) : 1.0;
      signal[i] *= w;
    }

    // === Hand-rolled DFT (real input, only positive freqs needed). ===
    // ~30 lines, O(n²) but n≤90 so fast enough.
    const halfN = Math.floor(n / 2) + 1;
    const magnitude = new Float64Array(halfN);
    for (let kIdx = 0; kIdx < halfN; kIdx++) {
      let re = 0;
      let im = 0;
      for (let t = 0; t < n; t++) {
        const angle = (-2 * Math.PI * kIdx * t) / n;
        re += signal[t] * Math.cos(angle);
        im += signal[t] * Math.sin(angle);
      }
      magnitude[kIdx] = Math.sqrt(re * re + im * im);
    }
    // Frequency axis: freqs[k] = k * fps / n.
    const dF = this.fps / n;

    let tremorSum = 0;
    let tremorCount = 0;
    let totalSum = 0;
    let totalCount = 0;
    for (let kIdx = 0; kIdx < halfN; kIdx++) {
      const f = kIdx * dF;
      if (f >= TREMOR_LOW_HZ && f <= TREMOR_HIGH_HZ) {
        tremorSum += magnitude[kIdx];
        tremorCount += 1;
      }
      if (f > 0.5) {
        totalSum += magnitude[kIdx];
        totalCount += 1;
      }
    }
    if (tremorCount === 0) return 0;
    const tremorPower = tremorSum / tremorCount;
    const totalPower = totalCount > 0 ? totalSum / totalCount : 1e-6;
    return tremorPower / Math.max(totalPower, 1e-6);
  }

  reset(): void {
    this.states.clear();
    this.frameTimes = [];
    this.currentLandmarks = null;
  }
}

/**
 * Centered moving average with mode='same' semantics:
 * out[i] = mean(signal[i-k/2 .. i+k/2]), clamping window at endpoints.
 */
function movingAverageSame(
  signal: ArrayLike<number>,
  k: number,
): Float64Array {
  const n = signal.length;
  const out = new Float64Array(n);
  const half = Math.floor(k / 2);
  // Naive O(nk). k≤15, n≤90 — < 1k mults.
  for (let i = 0; i < n; i++) {
    let s = 0;
    let c = 0;
    for (let j = i - half; j <= i + half; j++) {
      if (j >= 0 && j < n) {
        s += signal[j];
        c += 1;
      }
    }
    out[i] = c > 0 ? s / c : 0;
  }
  return out;
}

function round(x: number, digits: number): number {
  const k = Math.pow(10, digits);
  return Math.round(x * k) / k;
}
